arctic/store/version_store.py

```diff
@@ -576,6 +575,135 @@ def write(self, symbol, data, metadata=None, prune_previous_version=True, **kwar
+        version = {k: previous_version[k] for k in previous_version.keys()}
+        version['_id'] = bson.ObjectId()
+        version['version'] = new_version_num
+        version['metadata'] = metadata
+        version['base_version_id'] = previous_version.get('base_version_id', previous_version['_id'])
+
+        return self._add_new_version_using_reference(symbol, version, previous_version, prune_previous_version)
+
+    def restore_version(self, symbol, as_of, prune_previous_version=True, allow_secondary=None):
+        """
+        Restore the specified 'symbol' data and metadata to the state of a given version/snapshot/date.
+        Returns a VersionedItem object only with a metadata element.
+        Fast operation: Zero data/segment read/write operations.
+
+        Parameters
+        ----------
+        symbol : `str`
+            symbol name for the item
+        as_of : `str` or `int` or `datetime.datetime`
+            Return the data as it was as_of the point in time.
+            `int` : specific version number
+            `str` : snapshot name which contains the version
+            `datetime.datetime` : the version of the data that existed as_of the requested point in time
+        prune_previous_version : `bool`
+            Removes previous (non-snapshotted) versions from the database.
+            Default: True
+        allow_secondary : `bool` or `None`
+            Override the default behavior for allowing reads from secondary members of a cluster:
+            `None` : use the settings from the top-level `Arctic` object used to query this version store.
+            `True` : allow reads from secondary members
+            `False` : only allow reads from primary members
+
+        Returns
+        -------
+        `VersionedItem`
+            VersionedItem named tuple containing the metadata of the written symbol's version document in the store.
+        """
+        # if version/snapshot/data supplied in "as_of" does not exist, will fail fast with NoDataFoundException
+        version_to_restore = mongo_retry(self._read_metadata)(symbol, as_of=as_of,
+                                                              read_preference=self._read_preference(allow_secondary))
+
+        # Reaching here means that data and/or metadata exist and we are all set to update the metadata
+        new_version_num = self._version_nums.find_one_and_update({'symbol': symbol},
+                                                             {'$inc': {'version': 1}},
+                                                             upsert=True, new=True)['version']
+
+        # Create a new version entry, cloning the one supplied from the user
+        version = {k: version_to_restore[k] for k in version_to_restore.keys()}
+        version['_id'] = bson.ObjectId()
+        version['version'] = new_version_num
+        version['base_version_id'] = version_to_restore.get('base_version_id', version_to_restore['_id'])
```

Comment:
In terms of testing - I think it would be worth testing the GC at this point... Up till now there's only ever been one `base_version_id` in flight as it's only used during `append`.  After this change version history can 'fork'

Reply:
yes, very good point, I actually tested this manually and will add integration test.
The GC will preserve both the base_verserion_id and the new version document which we created and references it.

I chose this for performance, so that we don't have to update at all the segments by adding a new parent id in the parent set.

Reply:
Yep - the implementation looks good. Just wanted to check you'd thought about this :)